src/common/HordeAd/Core/SizedIndex.hs

```diff
@@ -0,0 +1,243 @@
+
+takeIndex :: forall m n i. KnownNat m
+          => Index (m + n) i -> Index m i
+takeIndex (Index ix) = Index $ takeSized ix
+
+dropIndex :: forall m n i. KnownNat m
+          => Index (m + n) i -> Index n i
+dropIndex (Index ix) = Index $ dropSized ix
+
+unsnocIndex :: Index (1 + n) i -> (Index n i, i)
+unsnocIndex (Index ix) = first Index $ unsnocSized ix
+
+lastIndex :: Index (1 + n) i -> i
+lastIndex (Index ix) = lastSized ix
+
+initIndex :: Index (1 + n) i -> Index n i
+initIndex (Index ix) = Index $ initSized ix
+
+permutePrefixIndex :: forall n i. KnownNat n
+                   => Permutation -> Index n i -> Index n i
+permutePrefixIndex p (Index ix) = Index $ permutePrefixSized p ix
+
+{-
+-- Look Ma, no unsafeCoerce! But it compiles only with GHC >= 9.2,
+-- so let's switch to it once we stop using 8.10 and 9.0.
+listToIndex :: forall n. KnownNat n => [Int] -> Index n Int
+listToIndex []
+  | Just Refl <- sameNat (Proxy @n) (Proxy @0) = ZI
+  | otherwise = error "listToIndex: list too short"
+listToIndex (i : is)
+  -- What we really need here to make the types check out is a <= check.
+  | EQI <- cmpNat (Proxy @1) (Proxy @n) =
+      let sh = listToIndex @(n - 1) is
+      in i :. sh
+  | LTI <- cmpNat (Proxy @1) (Proxy @n) =
+      let sh = listToIndexProxy @(n - 1) is
+      in i :. sh
+  | otherwise =
+      error "listToIndex: list too long"
+-}
+
+listToIndex :: forall n i. KnownNat n => [i] -> Index n i
+listToIndex = Index . listToSized
+
+indexToList :: Index n i -> [i]
+indexToList (Index l) = sizedListToList l
+
+
+-- * Tensor shapes as fully encapsulated sized lists, with operations
+
+-- | The shape of an n-dimensional array represented as a sized list.
+-- The order of dimensions corresponds to that in @Index@.
+newtype Shape n i = Shape (SizedList n i)
+  deriving Show
+
+pattern ZS :: forall n i. () => n ~ 0 => Shape n i
+pattern ZS = Shape Z
+
+infixr 3 :$
+pattern (:$) :: forall n1 i. forall n. (1 + n) ~ n1
+             => i -> Shape n i -> Shape n1 i
+pattern i :$ sh <- (unconsShape -> Just (UnconsShapeRes sh i Dict))
+  where i :$ (Shape sh) = Shape (i ::: sh)
+{-# COMPLETE ZS, (:$) #-}
+
+data UnconsShapeRes i n1 =
+  forall n. UnconsShapeRes (Shape n i) i (Dict (n1 ~ (1 + n)))
+unconsShape :: Shape n1 i -> Maybe (UnconsShapeRes i n1)
+unconsShape (Shape sh) = case sh of
+  i ::: sh' -> Just (UnconsShapeRes (Shape sh') i Dict)
+  Z -> Nothing
+
+deriving stock instance Functor (Shape n)
+
+instance KnownNat n => Foldable (Shape n) where
+  foldr f z l = foldr f z (shapeToList l)
+
+instance KnownNat n => IsList (Shape n i) where
+  type Item (Shape n i) = i
+  fromList = listShapeToShape
+  toList = shapeToList
+
+singletonShape :: i -> Shape 1 i
+singletonShape = Shape . singletonSized
+
+tailShape :: Shape (1 + n) i -> Shape n i
+tailShape (Shape ix) = Shape $ tailSized ix
+
+takeShape :: forall m n i. KnownNat m
+          => Shape (m + n) i -> Shape m i
+takeShape (Shape ix) = Shape $ takeSized ix
```

Comment:
In my code it's always `m`, so it's easier to compare. But in general, it may be anything, so `len` is neutral and descriptive in the module we aspire to include in `base`.